pennylane/_qubit_device.py

```diff
@@ -688,3 +689,110 @@ def sample(self, observable):
         unraveled_indices = [2] * len(device_wires)
         indices = np.ravel_multi_index(samples.T, unraveled_indices)
         return observable.eigvals[indices]
+
+    def adjoint_jacobian(self, tape):
+        """Implements the method outlined in https://arxiv.org/abs/2009.02823 to calculate the
+        Jacobian."""
+
+        for m in tape.measurements:
+            if m.obs is None:
+                raise ValueError(f"Adjoint differentiation method does not support measurement {m}")
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        # Perform the forward pass
+        self.reset()
+
+        # Consider using caching and calling lower-level functionality. We just need the state
+        # without postprocessing https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
+        self.execute(tape)
+
+        phi = self._reshape(self.state, [2] * self.num_wires)
+
+        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+
+        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
+                    ops = op.decomposition(*op.parameters, wires=op.wires)
+                    expanded_ops.extend(reversed(ops))
+                else:
+                    raise QuantumFunctionError(
+                        f"The {op.name} operation is not supported using "
+                        'the "adjoint" differentiation method'
+                    )
+            else:
+                expanded_ops.append(op)
+
+        expanded_ops = [o for o in expanded_ops if not o.name in ("QubitStateVector", "BasisState")]
+        dot_product_real = lambda a, b: self._real(math.sum(self._conj(a) * b))
+
+        param_number = len(tape._par_info) - 1
+        trainable_param_number = len(tape.trainable_params) - 1
+        for op in expanded_ops:
+
+            if op.grad_method and param_number in tape.trainable_params:
```

Comment:
```suggestion
            if (op.grad_method is not None) and (param_number in tape.trainable_params):
```

Something I learned recently.  Just because you *can* write it the way you did, doesn't mean it is the recommended way.

https://legacy.python.org/dev/peps/pep-0008/#programming-recommendations

Reply:
:100:
In particular, because so many things default to `False`, not using `is not None` is actually dangerous (we've actually had this gotcha several times in PL).
e.g., you could be using `None` to represent "no data", but actual data of the form `0`, `0.0` (i.e., with a numerical value which coincidentally has a boolean value of False) will cause an unintended conditional branch to be taken

Reply:
Great, thanks for the useful advice! I've updated this.